File: api/Dtos/PagedResult.cs
namespace CampusFacilities.Api.Dtos;

/// <summary>
/// One page of a list endpoint. Not a Result&lt;T&gt; wrapper — it carries no success or
/// error information, only the paging counters a client needs to render "page 2 of 7".
/// </summary>
public record PagedResult<T>(
    IReadOnlyList<T> Items,
    int Page,
    int PageSize,
    int TotalCount)
{
    public int TotalPages => PageSize <= 0 ? 0 : (int)Math.Ceiling(TotalCount / (double)PageSize);
}

File: mobile/lib/core/paged_result.dart
/// Mirrors the API's `PagedResult<T>` — the one pagination shape every list endpoint
/// returns. Only paging counters; it carries no success or error information.
class PagedResult<T> {
  const PagedResult({
    required this.items,
    required this.page,
    required this.pageSize,
    required this.totalCount,
    required this.totalPages,
  });

  final List<T> items;
  final int page;
  final int pageSize;
  final int totalCount;
  final int totalPages;

  bool get hasPrevious => page > 1;
  bool get hasNext => page < totalPages;

  factory PagedResult.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic> item) fromItem,
  ) {
    return PagedResult(
      items: (json['items'] as List<dynamic>)
          .map((item) => fromItem(item as Map<String, dynamic>))
          .toList(growable: false),
      page: json['page'] as int,
      pageSize: json['pageSize'] as int,
      totalCount: json['totalCount'] as int,
      totalPages: json['totalPages'] as int,
    );
  }
}
